pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java

```diff
@@ -0,0 +1,594 @@
+      } else {
+        byte[] valueBytes = StringUtil.encodeUtf8(value);
+        dataOutputStream.writeInt(valueBytes.length);
+        dataOutputStream.write(valueBytes);
+      }
+    }
+
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  private Map<MetadataKeys, String> deserializeMetadata(byte[] bytes)
+      throws IOException {
+    try (ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(bytes);
+        DataInputStream dataInputStream = new DataInputStream(byteArrayInputStream)) {
+      int numEntries = dataInputStream.readInt();
+      Map<MetadataKeys, String> metadata = new HashMap<>();
+      for (int i = 0; i < numEntries; i++) {
+        int keyId = dataInputStream.readInt();
+        Optional<MetadataKeys> opt = MetadataKeys.getByOrdinal(keyId);
+        // Ignore unknown keys.
+        if (!opt.isPresent()) {
+          continue;
+        }
+        MetadataKeys key = opt.get();
+        if (MetadataKeys.isIntValueMetadataKey(key)) {
+          String value = String.valueOf(decodeInt(dataInputStream));
+          metadata.put(key, value);
+        } else if (MetadataKeys.isLongValueMetadataKey(key)) {
+          String value = String.valueOf(decodeLong(dataInputStream));
+          metadata.put(key, value);
+        } else {
+          String value = String.valueOf(decodeString(dataInputStream));
+          metadata.put(key, value);
+        }
+      }
+      return metadata;
+    }
+  }
+
+  private byte[] serializeExceptions()
+      throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
+
+    dataOutputStream.writeInt(_exceptions.size());
+
+    for (Map.Entry<Integer, String> entry : _exceptions.entrySet()) {
+      int key = entry.getKey();
+      String value = entry.getValue();
+      byte[] valueBytes = StringUtil.encodeUtf8(value);
+      dataOutputStream.writeInt(key);
+      dataOutputStream.writeInt(valueBytes.length);
+      dataOutputStream.write(valueBytes);
+    }
+
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  private Map<Integer, String> deserializeExceptions(byte[] bytes)
+      throws IOException {
+    try (ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(bytes);
+        DataInputStream dataInputStream = new DataInputStream(byteArrayInputStream)) {
+      int numExceptions = dataInputStream.readInt();
+      Map<Integer, String> exceptions = new HashMap<>(numExceptions);
+      for (int i = 0; i < numExceptions; i++) {
+        int errCode = dataInputStream.readInt();
+        String errMessage = decodeString(dataInputStream);
+        exceptions.put(errCode, errMessage);
+      }
+      return exceptions;
+    }
+  }
+
+  @Override
+  public Map<String, String> getMetadata() {
+    return _metadataV2;
+  }
+
+  @Override
+  public DataSchema getDataSchema() {
+    return _dataSchema;
+  }
+
+  @Override
```

Comment:
Can we create a base class and move all of the following APIs to the base class which can be extended by both DataTableImplV2 and DataTableImplV3 ?